Source/WebCore/css/process-css-pseudo-selectors.py

```diff
@@ -0,0 +1,435 @@
+        if generator_type == GeneratorTypes['PSEUDO_CLASS_AND_COMPATIBILITY']:
+            writer.newline()
+            writer.write('#include "CSSParserSelector.h"')
+
+    def write_ignore_implicit_fallthrough(self, writer):
+        writer.newline()
+        writer.write('IGNORE_WARNINGS_BEGIN("implicit-fallthrough")')
+
+    def write_define_register(self, writer):
+        writer.newline()
+        writer.write('// Older versions of gperf like to use the `register` keyword.')
+        writer.write('#define register')
+
+    def open_namespace_webcore(self, writer):
+        writer.newline()
+        writer.write('namespace WebCore {')
+
+    def write_entry_struct_definition(self, writer, generator_type):
+        if generator_type == GeneratorTypes['PSEUDO_CLASS_AND_COMPATIBILITY']:
+            writer.write_block("""
+                struct SelectorPseudoClassOrCompatibilityPseudoElementEntry {
+                    const char* name;
+                    PseudoClassOrCompatibilityPseudoElement pseudoTypes;
+                };""")
+        else:
+            writer.write_block("""
+                struct SelectorPseudoTypeEntry {
+                    const char* name;
+                    std::optional<CSSSelector::PseudoElement> type;
+                };""")
+
+    def write_gperf_mapping(self, writer, generator_type):
+        writer.write_block("""
+            %struct-type
+            %define initializer-suffix ,{0}
+            %define class-name {1}
+            %omit-struct-type
+            %language=C++
+            %readonly-tables
+            %global-table
+            %ignore-case
+            %compare-strncmp
+            %enum
+
+            struct {2};
+            """.format(
+            '{std::nullopt,std::nullopt}' if generator_type == GeneratorTypes['PSEUDO_CLASS_AND_COMPATIBILITY'] else 'std::nullopt',
+            'SelectorPseudoClassAndCompatibilityElementMapHash' if generator_type == GeneratorTypes['PSEUDO_CLASS_AND_COMPATIBILITY'] else 'SelectorPseudoElementMapHash',
+            'SelectorPseudoClassOrCompatibilityPseudoElementEntry' if generator_type == GeneratorTypes['PSEUDO_CLASS_AND_COMPATIBILITY'] else 'SelectorPseudoTypeEntry',
+        ))
+
+        writer.write('%%')
+
+        for key in self.mapping:
+            value = self.mapping[key]
+            if generator_type == GeneratorTypes['PSEUDO_CLASS_AND_COMPATIBILITY']:
+                writer.write('"{}", {{{}, {}}}'.format(key, value[0], value[1]))
+            else:
+                writer.write('"{}", {}'.format(key, value))
+
+        writer.write('%%')
+
+    def write_parsing_function_definitions_for_pseudo_class(self, writer):
+        longest_keyword_length = len(max(self.mapping, key=len))
+        writer.write_block("""
+        static inline const SelectorPseudoClassOrCompatibilityPseudoElementEntry* parsePseudoClassAndCompatibilityElementString(const LChar* characters, unsigned length)
```

Comment:
It’s not a great pattern to have so much C++ code here in the python source that is unconditional and doesn’t vary. Ideally more of the code would be in `.h` and `.cpp` files. For example we can just emit a constexpr that is the size of the maximum keyword length and have the function include that generated header but not itself be generated.

Reply:
I can look into doing this as a follow up, I just left status quo since I'm not too familiar with how those gperf files work in general.

Reply:
Filed bug https://bugs.webkit.org/show_bug.cgi?id=266960